main/autogen/autogen.cc

```diff
@@ -663,15 +664,116 @@ string ParsedFile::toMsgpack(core::Context ctx, int version) {
             continue;
         }
         auto names = showFullName(ctx, def.id);
-        out.emplace_back(fmt::format("{}", fmt::map_join(names, "::", nameToString)));
+        out.emplace_back(fmt::format("{}", fmt::map_join(names, "::", [&ctx](const core::NameRef &nm) -> string {
+                                         return nm.data(ctx)->show(ctx);
+                                     })));
+    }
+
+    return out;
+}
+
+optional<AutogenSubclassMap> ParsedFile::listAllSubclasses(const core::Context ctx,
+                                                           const vector<string> &absolutePathsToIgnore,
+                                                           const vector<string> &relativePathsToIgnore) {
+    // We prepend "/" to `path` to mimic how `isFileIgnored` gets called elsewhere
+    if (sorbet::FileOps::isFileIgnored("", fmt::format("/{}", path), absolutePathsToIgnore, relativePathsToIgnore)) {
+        return nullopt;
+    }
+
+    AutogenSubclassMap out;
+
+    for (const Reference &ref : refs) {
+        DefinitionRef defn = ref.parent_of;
+
+        if (!defn.exists()) {
+            continue;
+        }
+
+        // Get fully-qualified parent name as string
+        string parentName =
+            fmt::format("{}", fmt::map_join(ref.resolved, "::", [&ctx](const core::NameRef &nm) -> string {
+                            return nm.data(ctx)->show(ctx);
+                        }));
+
+        // Add child class to the set identified by its parent
+        string childName =
+            fmt::format("{}", fmt::map_join(showFullName(ctx, defn), "::", [&ctx](const core::NameRef &nm) -> string {
+                            return nm.data(ctx)->show(ctx);
+                        }));
+
+        out[parentName].insert(make_pair(childName, defn.data(*this).type));
+    }
+
+    return out;
+}
+
+// Generate all descendants of a parent class
+// Recursively walks `childMap`, which stores the IMMEDIATE children of subclassed class.
+void descendantsOf(const AutogenSubclassMap &childMap, const string &parentName, AutogenSubclassSet &out) {
+    if (!childMap.count(parentName)) {
+        return;
+    }
+    const AutogenSubclassSet &children = childMap.at(parentName);
+
+    out.insert(children.begin(), children.end());
+    for (const AutogenSubclassEntry &child : children) {
+        descendantsOf(childMap, child.first, out);
     }
 }
 
+void maybeInsertChild(const string &parentName, const AutogenSubclassSet &children, AutogenSubclassMap &out) {
+    if (parentName.empty()) {
+        // Child < NonexistentParent
+        return;
+    }
+    out[parentName].insert(children.begin(), children.end());
+}
+
+// Manually patch the child map to account for inheritance that happens at runtime `self.included`
+// Please do not add to this list.
+void patchChildMap(AutogenSubclassMap &childMap) {
+    childMap["Opus::SafeMachine"].insert(childMap["Opus::Risk::Model::Mixins::RiskSafeMachine"].begin(),
```

Comment:
Could these become part of rbis?

Reply:
@DarkDimius could you elaborate bit on what you mean?

Reply:
IRL'd, I'll take care of this in a follow-on PR.